src/go/privileged-service/pkg/port/portproxy.go

```diff
@@ -17,27 +17,69 @@ limitations under the License.
 package port
 
 import (
+	"crypto/md5"
+	"encoding/json"
+	"errors"
 	"fmt"
 	"net"
+	"sort"
+	"sync"
 
+	"github.com/docker/go-connections/nat"
 	"github.com/rancher-sandbox/rancher-desktop-agent/pkg/types"
 	"github.com/rancher-sandbox/rancher-desktop/src/go/privileged-service/pkg/command"
 )
 
 const netsh = "netsh"
 
-func execProxy(portMapping types.PortMapping) error {
+var ErrPortProxy = errors.New("error from PortProxy")
+
+type portProxy struct {
+	PortMap      nat.PortMap
+	ConnectAddrs []types.ConnectAddrs
+}
+
+type proxy struct {
+	portMappings map[string]portProxy
+	mutex        sync.Mutex
+}
+
+func newProxy() *proxy {
+	return &proxy{
+		portMappings: make(map[string]portProxy),
+	}
+}
+
+func (p *proxy) exec(portMapping types.PortMapping) error {
+	port := portProxy{
+		PortMap:      portMapping.Ports,
+		ConnectAddrs: portMapping.ConnectAddrs,
+	}
 	if portMapping.Remove {
-		return deleteProxy(portMapping)
+		return p.delete(port)
 	}
-	return addProxy(portMapping)
+	return p.add(port)
+}
 
+func (p *proxy) removeAll() error {
+	errs := make([]error, 0)
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	for _, proxy := range p.portMappings {
+		if err := p.delete(proxy); err != nil {
+			errs = append(errs, fmt.Errorf("deleting portproxy: %+v faile: %w", proxy, err))
```

Comment:
For better formatting, consider something like https://pkg.go.dev/github.com/hashicorp/go-multierror. (I'm fine merging this without it, though.)

Reply:
I'm starting to get concerned about our binary size on Windows. I would prefer if we can avoid bringing in extra dependencies.

Reply:
What is the concern about size? Is adding another Go package really a problem?

Reply:
The concern is go by default statically link its dependencies. Although, a single dependency may not seem like much what that dependency depends on could be kind of exponential.